[Liquid Glass] [macOS] Add a way to extend the height of the scroll pocket

whsieh opened the pull request:
#### 8ac8e902cceaddc21445b0837575fa16a494d5a1
<pre>
[Liquid Glass] [macOS] Add a way to extend the height of the scroll pocket
<a href="https://bugs.webkit.org/show_bug.cgi?id=297106">https://bugs.webkit.org/show_bug.cgi?id=297106</a>
<a href="https://rdar.apple.com/157815756">rdar://157815756</a>

Reviewed by Aditya Keerthi.

Add a new macOS-specific property, `-_setOverflowHeightForTopScrollEdgeEffect:`, to allow Safari to
extend the height of the web view&apos;s scroll pocket upwards by a specified amount. See below for more
details.

* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:

Add `overflowHeightForTopScrollEdgeEffect` to WebPage creation parameters, so that newly created web
views will start with the correct height.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView scrollViewDrawsMagicPocket]):

Return true if either the top obscured content inset or the additional overflow height is nonzero.

(-[WKWebView _overflowHeightForTopScrollEdgeEffect]):
(-[WKWebView _setOverflowHeightForTopScrollEdgeEffect:]):
* Source/WebKit/UIProcess/Inspector/mac/WKInspectorViewController.mm:
(-[WKInspectorViewController didAttachOrDetach]):

Also propagate this SPI over to the attached inspector web view, so that docked inspector views will
also get the extra pocket height.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::setOverflowHeightForTopScrollEdgeEffect):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::updateScrollPocket):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::sidesRequiringFixedContainerEdges const):

Make sure we sample for the top fixed container, in the case where a nonzero overflow top pocket
height is set.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ObscuredContentInsets.mm:
(TestWebKitAPI::TEST(ObscuredContentInsets, OverflowHeightForTopScrollEdgeEffect)):

Add an API test to exercise this new property.

Canonical link: <a href="https://commits.webkit.org/298394@main">https://commits.webkit.org/298394@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cd9f37e46b86c3c1b3c418d75f423118f584809a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115372 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35066 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25568 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121462 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65950 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c6cbdd3f-d8d5-4786-8116-d8d688bb1d28) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117261 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35725 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43638 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87654 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42356 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2a06f145-3a48-487a-b491-6da8303b5069) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118320 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28492 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103566 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68048 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27650 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21688 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65121 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97880 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21801 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124627 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42311 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31692 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96436 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42679 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99754 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96223 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24483 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41455 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19312 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/38230 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42194 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47747 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41697 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45025 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43417 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
```diff
@@ -1609,6 +1609,23 @@ - (NSEdgeInsets)_obscuredContentInsets
+        return;
+
+    _page->setOverflowHeightForTopScrollEdgeEffect(height);
+    _impl->updateScrollPocket();
```
I think you need a `ENABLE(CONTENT_INSET_BACKGROUND_FILL)` here?

whsieh replied:
Whoops, yes!

whsieh committed:
[Liquid Glass] [macOS] Add a way to extend the height of the scroll pocket
https://bugs.webkit.org/show_bug.cgi?id=297106
rdar://157815756

Reviewed by Aditya Keerthi.

Add a new macOS-specific property, `-_setOverflowHeightForTopScrollEdgeEffect:`, to allow Safari to
extend the height of the web view's scroll pocket upwards by a specified amount. See below for more
details.

* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:

Add `overflowHeightForTopScrollEdgeEffect` to WebPage creation parameters, so that newly created web
views will start with the correct height.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView scrollViewDrawsMagicPocket]):

Return true if either the top obscured content inset or the additional overflow height is nonzero.

(-[WKWebView _overflowHeightForTopScrollEdgeEffect]):
(-[WKWebView _setOverflowHeightForTopScrollEdgeEffect:]):
* Source/WebKit/UIProcess/Inspector/mac/WKInspectorViewController.mm:
(-[WKInspectorViewController didAttachOrDetach]):

Also propagate this SPI over to the attached inspector web view, so that docked inspector views will
also get the extra pocket height.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::setOverflowHeightForTopScrollEdgeEffect):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::updateScrollPocket):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::sidesRequiringFixedContainerEdges const):

Make sure we sample for the top fixed container, in the case where a nonzero overflow top pocket
height is set.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ObscuredContentInsets.mm:
(TestWebKitAPI::TEST(ObscuredContentInsets, OverflowHeightForTopScrollEdgeEffect)):

Add an API test to exercise this new property.

Canonical link: https://commits.webkit.org/298394@main
